pep-0702.rst

```diff
@@ -0,0 +1,278 @@
+work as is, not with potential future changes. However, the analysis that type
+checkers perform on code to find type errors is very similar to the analysis
+that would be needed to detect usage of many deprecations. Therefore, type
+checkers are well placed to find and report deprecations.
+
+Other languages already have similar functionality:
+
+* GCC supports a ``deprecated`` `attribute <https://gcc.gnu.org/onlinedocs/gcc-3.1.1/gcc/Type-Attributes.html>`__
+  on function declarations. This powers CPython's ``Py_DEPRECATED`` macro.
+* GraphQL `supports <https://spec.graphql.org/June2018/#sec-Field-Deprecation>`__
+  marking fields as ``@deprecated``.
+* Kotlin `supports <https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-deprecated/>`__
+  a ``Deprecated`` annotation.
+* Scala `supports <https://www.scala-lang.org/api/2.12.5/scala/deprecated.html>`__
+  an ``@deprecated`` annotation.
+* Swift `supports <https://docs.swift.org/swift-book/ReferenceManual/Attributes.html>`__
+  using the ``@available`` attribute to mark APIs as deprecated.
+* TypeScript `uses <https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#deprecated>`__
+  the ``@deprecated`` JSDoc tag to issue a hint marking use of
+  deprecated functionality.
+
+Several users have requested support for such a feature:
+
+* `typing-sig thread <https://mail.python.org/archives/list/typing-sig@python.org/thread/E24WTMQUTGKPFKEXVCGGEFFMG7LDF3WT/>`__
+* `Pyright feature request <https://github.com/microsoft/pyright/discussions/2300>`__
+* `mypy feature request <https://github.com/python/mypy/issues/11439>`__
+
+
+Specification
+=============
+
+The new ``@typing.deprecated()`` decorator can be used on a class, function or method.
+With overloaded functions, the decorator may be applied to individual overloads,
+indicating that the particular overload is deprecated. The decorator may also be
+applied to the overload implementation function, indicating that the entire function
+is deprecated.
+
+The decorator takes a single argument of type ``str``, which is a message that should
+be shown by the type checker when it encounters a usage of the decorated object.
+The message must be a string literal or a string with a ``Literal`` type.
+The content of deprecation messages is up to the user, but it may include the version
+in which the deprecated object is to be removed, and information about suggested
+replacement APIs.
+
+Type checkers should produce a diagnostic whenever they encounter a usage of an
+object marked as deprecated. For deprecated overloads, this includes all calls
+that resolve to the deprecated overload.
+For deprecated classes and functions, this includes:
+
+* References through module, class, or instance attributes (``module.deprecated_object``,
+  ``module.SomeClass.deprecated_method``, ``module.SomeClass().deprecated_method()``)
+* Any usage of deprecated objects in their defining module
+  (``x = deprecated_object()`` in ``module.py``)
```

Comment:
This one could be problematic? *Internal* uses might need to be kept around, if the deprecated object returns something whose API is not deprecated. OTOH after
```py
deprecated_alias = deprecated_object
```
using `module.deprecated_alias` elsewhere should be considered deprecated.

Reply:
I'm not sure why internal uses are special; e.g. when the `assertEquals` method in unittest is deprecated, unittest itself shouldn't be using it either.